lighthouse-core/lib/dependency-graph/base-node.js

```diff
@@ -168,96 +168,79 @@ class BaseNode {
 
   /**
    * Clones the entire graph connected to this node filtered by the optional predicate. If a node is
-   * included by the predicate, all nodes along the paths between the two will be included. If the
-   * node that was called clone is not included in the resulting filtered graph, the method will throw.
+   * included by the predicate, all nodes along the paths between the node and the root will be included. If the
+   * node this was called on is not included in the resulting filtered graph, the method will throw.
    * @param {function(Node):boolean} [predicate]
    * @return {Node}
    */
   cloneWithRelationships(predicate) {
     const rootNode = this.getRootNode();
 
-    /** @type {function(Node): boolean} */
-    let shouldIncludeNode = () => true;
-    if (predicate) {
-      const idsToInclude = new Set();
-      rootNode.traverse(node => {
-        if (predicate(node)) {
-          node.traverse(
-            node => idsToInclude.add(node.id),
-            node => node._dependencies.filter(parent => !idsToInclude.has(parent))
-          );
-        }
-      });
-
-      shouldIncludeNode = node => idsToInclude.has(node.id);
-    }
+    /** @type {Map<string, Node>} */
+    const idsToIncludedClones = new Map();
 
-    const idToNodeMap = new Map();
-    rootNode.traverse(originalNode => {
-      if (!shouldIncludeNode(originalNode)) return;
-      const clonedNode = originalNode.cloneWithoutRelationships();
-      idToNodeMap.set(clonedNode.id, clonedNode);
+    // Walk down dependents.
+    rootNode.traverse(node => {
+      // Find new and includable nodes.
+      if (idsToIncludedClones.has(node.id)) return;
+      if (predicate && !predicate(node)) return;
+
+      // Walk back up dependencies, cloning all nodes in the path to this one.
```

Comment:
can we only do the reverse traversal when there's a predicate? 

that's one nice feature of the previous impl I wouldn't want to lose

Reply:
> can we only do the reverse traversal when there's a predicate?

do you mean something like

```js
// Walk down dependents.
rootNode.traverse(node => {
  if (idsToIncludedClones.has(node.id)) return;

  if (!predicate) {
    // Clone every node.
    idsToIncludedClones.set(node.id, node.cloneWithoutRelationships());
  } else if (predicate(node)) {
    // Walk back up dependencies, cloning all nodes in the path to this one.
    node.traverse(
      node => idsToIncludedClones.set(node.id, node.cloneWithoutRelationships()),
      // Dependencies already cloned have already cloned ancestors, so no need to visit again.
      node => node._dependencies.filter(parent => !idsToIncludedClones.has(parent.id))
    );
  }
});
```

Or other ideas for formatting to make it less dense looking?

Another option is to make a default always true predicate, so it's a little clearer.

Agreed it is a bit weird to do the reverse traversal in the no predicate case. My thinking before was that we only have a single call of `cloneWithRelationships` without a predicate, so it didn't seem to be a super important case to call out (and since the filter line makes there be no next nodes for every node, it doesn't do any significant extra work)

Reply:
that's exactly what I meant :)

we could do an early return in the `!predicate` case to avoid the nesting? but IMO this is fine

Reply:
done